examples/active_replication/flux_buffer.py

```diff
@@ -0,0 +1,54 @@
+
+class Buffer:
+    def __init__(self, n_dest):
+        self.n_dest = n_dest
+        self.queue = list()  # TODO(yika): add upper bound on buffer size
+
+    def put(self, data, sn, dest):
+        # Put input stream
+        if len(self.queue) > 0:
+            assert sn >= self.queue[-1][0]
+        self.queue.append((sn, data, dest))
+
+    def ack(self, sn, dest):
+        for i in range(len(self.queue)):
+            if self.queue[i][0] == sn and self.queue[i][2] == dest:
+                self.queue.pop(i)
+                return True
+        return False
+
+    def ack_all(self, dest):
+        i = 0
+        while i < len(self.queue):
+            if self.queue[i][2] == dest:
+                self.queue.pop(i)
```

Comment:
I see now why you're queuing a message for each replica. The approach I proposed above is not sufficient to implement this method. However, you can still avoid duplicating every data message in the queue by stroing (seq_num, data, list_of_non_acknowledge_replicas).

Reply:
this was my previous implementation, but then I realize the paper does it the current way so I figure I should follow it. Also it's more wordy to implement that because tuple is immutable so it's better to do only checking and dropping the whole tuple instead of modifying it

Reply:
Ok, you don't actually duplicate data because you store references in the queue. Then this is fine. It's just that the design is less brittle if you don't assume that replicas have ids starting from zero. 

For example, imagine a setup in which the data-flow runs for a long time, while nodes fail and come back. After every single failure, you would have to ensure that the replica that replaces the failed one gets the id of the failed one. If you don't make the assumption that ids are sequential, and starting from zero, you don't have to do anything (wrt to the data structure) when you bring up a new replica.

Reply:
Re immutable: Good point, however, the tuple stores a reference to the list. Hence, you don't modify the tuple, only the list it references to.

Ignore my comment on the same id.

Ok let's leave it as it is, but please add a comment to explain the alternative option.